_posts/2016-04-19-MIR.md

```diff
@@ -0,0 +1,725 @@
+---
+layout: post
+title: "Introducing MIR"
+author: Niko Matsakis
+description: "The shift to use MIR in the compiler should unlock many exciting improvements."
+---
+
+We are in the final stages of a grand transformation on the Rust
+compiler internals. Over the past year or so, we have been steadily
+working on a plan to change our internal compiler pipeline, as shown
+here:
+
+<center>
+![Compiler Flowchart][flow]
+</center>
+
+That is, we are introducing a new intermediate representation (IR) of
+your program that we call **MIR**: MIR stands for *middle* IR, because
+the MIR comes between the existing HIR ("high-level IR", roughly an
+[abstract syntax tree][ast]) and [LLVM][llvm] (the "low-level"
+IR). Previously, the "translation" phase in the compiler would convert
+from full-blown Rust into machine-code-like LLVM in one rather large
+step. But now, it will do its work in two phases, with a vastly
+simplified version of Rust -- MIR -- standing in the middle.
+
+If you're not a compiler enthusiast, this all might seem arcane and unlikely to
+affect you directly. But in reality, **MIR is the key to ticking off a number of
+our highest priorities** for Rust:
+
+-   **Faster compilation time**. We are working to make Rust's compilation
+    *incremental*, so that when you *re*-compile code, the compiler recomputes only
+    what it has to. MIR has been designed from the start with this use-case in mind,
+    so it's much easier for us to save and reload, even if other parts of the program
+    have changed in the meantime.
+
+    MIR also provides a foundation for more efficient data structures and removal
+    of redundant work in the compiler, both of which should speed up compilation
+    across the board.
+
+-   **Faster execution time**. You may have noticed that in the new compiler
+    pipeline, *optimization* appears twice. That's no accident: previously, the
+    compiler relied solely on LLVM to perform optimizations, but with MIR, we can
+    do some *Rust-specific* optimizations before ever hitting LLVM -- or, for that
+    matter, before monomorphizing code. Rust's rich type system should provide
+    fertile ground for going beyond LLVM's optimizations.
+
+    In addition, MIR will uncork some longstanding performance improvements
+    to the code Rust generates, like ["non-zeroing" drop][rfc320].
```

Comment:
Given that we dive into non-zeroing drop later on, it would be nice if you could explain what it is up front (around here), rather than just linking.